sympy/functions/combinatorial/numbers.py

```diff
@@ -996,6 +997,174 @@ def _eval_evalf(self, prec):
+                  '%6s' % str(is_prime(c))
+
+        start = 100
+        end = 10000
+        print "\nCarmichael numbers between " + str(start) + " and " + str(end) + ":"
+        print str(find_carmichaels_in_range(start, end))
+    
+     Output
+    ========
+     >>> 
+  ================ RESTART: C:/Python27/sympycarmichael.py ===============
+    First 10 Carmichael Numbers:
+    1) 561
+    2) 1105
+    3) 1729
+    4) 2465
+    5) 2821
+    6) 6601
+    7) 8911
+    8) 10585
+    9) 15841
+    10) 29341
+    [561, 1105, 1729, 2465, 2821, 6601, 8911, 10585, 15841, 29341]
+
+    ---Fermat Prime Tests---
+    Carmichael Miller Rabin Truth
+           561| False
+          1105| False
+          1729| False
+          2465| False
+          2821| False
+          6601| False
+          8911| False
+         10585| False
+         15841| False
+         29341| False
+
+    Carmichael numbers between 100 and 10000:
+    Set([128, 512, 8192, 2821, 2048, 1729, 6601, 8911, 1105, 561, 2465])
+
+    """
+    knownPrimes = Set()
+    # returns true if n is a perfect square
+    def is_perfect_square(apositiveint):
+        x = apositiveint // 2
+        seen = Set([x])
+        while x * x != apositiveint:
+            x = (x + (apositiveint // x)) // 2
+            if x in seen: return False
+            seen.add(x)
+        return True
+
+    # returns true if p divides n
+    def divides(p, n):
+        return n%p == 0
+
+    # returns true if n is prime  
+    def is_prime(n):
+        return isprime(n)
+
+        global knownPrimes
+        if n in knownPrimes:
+            return True
+        else:
+            for i in range(3, n/2, 2):
+                if not (n%i):
+                    return False
+
+        knownPrimes.add(n)
+        return True
+    # returns true if n is a Carmichael number
+    def is_carmichael(n):
+        if n==1:
+            return False
+        if is_prime(n):
+            return False
+
+        divisors = Set([1, n])
+        # get divisors
+        for i in xrange(3, n/2+1, 2):
+            if n%i == 0:
+                divisors.add(i)
+
+        for i in divisors:
+            if is_perfect_square(i) and i != 1:
+                return False
+            if is_prime(i):
+                if not divides(i-1, n-1):
+                    return False
+
+        return True
+
+    # returns a set object of Carmichaels in [x, y],
+    # where x and y are non-negative integers and x <= y
+    def find_carmichaels_in_range(x, y):
+        return Set([i for i in xrange(x, y) if is_carmichael(i)])
+
+    # returns a set of the first n Carmichael numbers
+    # beginning its search from 0
+    # As the numbers are added to that set they are printed out
+    def find_first_n_carmichaels(n):
+        i = 1
+        carmichaels = Set()
+        while len(carmichaels) < n:
+            if is_carmichael(i):
+                carmichaels.add(i)
+                print str(len(carmichaels)) + ") " + str(i)
```

Comment:
The numbers need not be printed while they are being stored.
Also, what is the purpose of using a `Set()`?